x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardChangesAction.java

```diff
@@ -297,12 +297,13 @@ protected Response newResponse() {
         if (indexShard.state() != IndexShardState.STARTED) {
             throw new IndexShardNotStartedException(indexShard.shardId(), indexShard.state());
         }
-        if (fromSeqNo > indexShard.getGlobalCheckpoint()) {
```

Comment:
Question: so `indexShard.getGlobalCheckpoint()` may return a lower seqno than acquired from `indexShard.seqNoStats().getGlobalCheckpoint()`? I always assumed that the seqno acquired from `IndexShard` could not go backwards.

Reply:
The global checkpoint in the stats object and directly from the index shard are sourced from the same place, the replication tracker. The problem here, as I understand it, is that the global checkpoint could have advanced after capturing the stats. Here is what can happen then:
 - suppose that `fromSeqNo` is 17
 - suppose that the global checkpoint in the stats instance is 16
 - suppose that the global checkpoint advances to 17 after the stats object is captured
 - the `fromSeqNo > indexShard.getGlobalCheckpoint()` check will fail (because of the advance), meaning that we skip returning an empty operations response
 - we then calculate `toSeqNo = Math.min(globalCheckpoint, (fromSeqNo + maxOperationCount) - 1)` where `globalCheckpoint` is from the stats instance; this would give `toSeqNo == 16`
 - now we have `[fromSeqNo, toSeqNo] == [17, 16]` which produces the invalid range error message

This all happened because we allowed the global checkpoint advancing to become visible to this logic. Had we reused `globalCheckpoint` from the stats object then `fromSeqNo > globalCheckpoint` would have succeeded and we would have returned an empty operations response.

Reply:
@jasontedor Thanks for the explanation!

Reply:
Great explanation!